### DISCUSSION TITLE
Sensor Calibration Methodology Ideas

### DISCUSSION THREAD

**Discussion by DrewBrandt** [Terrapin-Rocket-Team/SRAD-Avionics#109] [Ideas]
This is how I would calibrate each sensor to avoid drift while sitting on the pad:

- Note that we already did a slightly flawed version of this at SAC24 and URRG24

### Barometer:
Continuously reset ground level to the current altitude while a launch is not detected

### IMU:
Continuously calculate a drift offset for gyros and accels and apply that offset to the reported measurement while launch is not detected.

### GPS:
Continuously reset starting position while launch is not detected

## Method:

All of these are basically identical in practice. The sensors are measuring some value while at rest on the pad, and that value needs to be removed from the value given to the state system to get the delta, which is our real position/speed/accel.

I would keep a record of the last 20 or so measurements and use the first 10 of thsoe to average out a value as our "offset". Ignore the most recent ten readings (assuming 10hz measurements) to avoid taking in launch data and using it as the baseline. It is assumed that the rocket will detect launch and stop calculating these offsets within one second of it moving from rest.

**Comment by varun-un** [comment]
The one thing about this method is I feel they assign linear or constant biases to the sensor which in the case of a more dynamic sensor like the IMU may happen less, so that may affect this way of zeroing it. This may need to be tested and confirmed to be valid at more obtuse orientations and higher accels.

**Reply by DrewBrandt** [reply]
I agree, although I don't know of a way for us to test a known high g value.

**Comment by varun-un** [comment]
Okay, apparently (from my very brief and not thorough research), is that IMUs may have constant biases, which this method will work for, but they may also have temperature biases and bias instability. If we really feel that temperature bias is an issue, we can characterize that experimentally and have a very simple lookup table for bias corrections based off of a temperature sensor reading in the code. For bias instability, apparently what you can do is then [add bias terms to the state of the Kalman Filter](https://stackoverflow.com/questions/71270530/code-examples-for-imu-velocity-estimation/71295138#71295138). Since a KF essentially expects gaussian noise on sensor readings, it doesn't do too well to filter out those biases, but you can do some funky thing to add a bias term to your state that it'll then predict, and then your dynamic model will use this bias term to correct.

This may all be overkill though if we could just select a sensor with minimal bias to begin with. This is also maybe better for in flight as well
